app/front/scripts/controllers/main.js

```diff
@@ -0,0 +1,153 @@
+                  $scope.state.dimensions.current.groups.splice(index, 1);
+                }
+              } else {
+//              $scope.state.dimensions.current.groups.push(group);//babbage.ui doesn't support multy-drilldown
+                $scope.state.dimensions.current.groups = [group];
+              }
+              updateBabbage();
+            };
+            $scope.events.changeFilter = function (filter, value) {
+              $scope.state.dimensions.current.filters[filter] = value;
+              updateBabbage();
+            }
+            $scope.events.dropFilter = function (filter) {
+              delete $scope.state.dimensions.current.filters[filter];
+              updateBabbage();
+            };
+            $scope.events.setTab = function (aTab){
+              $scope.currentTab = aTab;
+            };
+            $scope.events.canBack = function (){
+              return HistoryService.canBack();
+            };
+            $scope.events.canForward = function (){
+              return HistoryService.canForward();
+            }
+            $scope.events.back = function (){
+              var history = HistoryService.back();
+              setState(history);
+            };
+            $scope.events.forward = function (){
+              var history = HistoryService.forward();
+              setState(history);
+            }
+          }
+
+          function setState(state) {
+            $scope.state = _.extend($scope.state, state);
+            $timeout(function(){
+              $scope.state.flag.renderingCharts = true;
+              $timeout(function() {
+                $scope.state.flag.renderingCharts = false;
+              })
+            });
+          }
+
+          function changePackage(packageName) {
+            $scope.currentTab = 'Treemap';
+            $scope.state.isPackageLoading = true;
+            $scope.state.availablePackages.current = packageName;
+
+            ApiService.getPackage(packageName).then(function (package) {
+              $scope.state.availablePackages.description = (package.description) ? package.description : package.title;
+            });
+
+            $scope.state.measures.current = '';
+            $scope.state.dimensions.current.groups = [];
+            $scope.state.dimensions.current.filters = {};
+
+            ApiService.getPackageModel(packageName).then(function (packageModel){
+              $scope.state.dimensions.items = packageModel.dimensions.items;
+              $scope.state.measures.items = packageModel.measures.items;
+              $scope.state.measures.current = packageModel.measures.current;
+              $scope.state.dimensions.current.groups = [(_.first(packageModel.dimensions.items)).key];
+
+            }).finally(function(){
+              $scope.state.isPackageLoading = false;
+              updateBabbage();
+            });
+          }
+
+          function updateBabbage(){
+            $timeout(function(){
```

Comment:
we use $timeout for some async operation. I this case, we should "refresh" babble.ui components (delete it and add again) because they don't watch for data changing.